src/dimension/mod.rs

```diff
@@ -394,6 +382,19 @@ fn to_abs_slice(axis_len: usize, slice: Slice) -> (usize, usize, isize) {
     (start, end, step)
 }
 
+/// The offset value of the data pointer relative to the actual pointer of the given
+/// dim and strides. The result is always <= 0.
+pub fn head_ptr_offset(dim: &[Ix], strides: &[Ix]) -> isize{
```

Comment:
Let's formulate this using a wording of "from.. to", then the sign of the pointer difference should be clear. For example, "this computes the offset **from** the first element in memory **to** the logically first element of the array".

I'd like the function name to be based on the same thinking.

Should this function not use a type parameter `D` so that it is special cased automatically for the most important dimensions (1 and 2)?

And missing space:

```suggestion
pub fn head_ptr_offset(dim: &[Ix], strides: &[Ix]) -> isize {
```

Reply:
I'm sorry I lacked some skills. Why change the parameter can provide a specific implementation for a 1d or 2d array? I hope there are examples for me to refer to.